File: 22-generate-parentheses/generate-parentheses.cpp
class Solution {
public:

    void backtrack(vector<string>& ans,string entry,int op,int cp)
    {
        if(op==0 && cp==0)
        {
            ans.push_back(entry);
            return;
        }
        if(op>0)
        {
            entry+="(";
            backtrack(ans,entry,op-1,cp);
            entry=entry.substr(0,entry.length()-1);
        }
        if(cp>op)
        {
            entry+=")";
            backtrack(ans,entry,op,cp-1);
            entry=entry.substr(0,entry.length()-1);
        }
    }

    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        string entry;
        backtrack(ans,entry,n,n);
        return ans;
    }
};

File: 22-generate-parentheses/generate-parentheses.java
class Solution {

    //recursive function
    public void helper(int open,int close,String curr,List<String> ans,int n) {
        //base case (ending condition)
        if(curr.length()==2*n)
        {
            ans.add(curr);
            return;
        }

       //open use
        if(open<n)
        {
            helper(open+1,close,curr+"(",ans,n);
        }

        //close use
        if(open>close)
        {
            helper(open,close+1,curr+")",ans,n);
        }
        
    }

    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<>();
        int open = 0;
        int close = 0;
        String curr = "";
        helper(open,close,curr,ans,n);
        return ans;
    }
}